### ISSUE TITLE
Add a conditional branching operator to `Mono`?

### ISSUE THREAD

**Issue by simonbasle** [reactor/reactor-core#1186]
Statically choosing between two path can be done with a classic imperative `if` statement. But reactively choosing a path from the result of a `Mono` could benefit from a dedicated operator. If the source is a `Mono<T>`, we'd probably want a `Predicate<T>` to choose the path. Ideally, if the source is a `Mono<Boolean>`, that could be considered as the "predicate".

I created a candidate implementation that represents what I have in mind:
```java

	private static final Predicate<Object> BOOLEAN_OR_THERE_PREDICATE = (Object o) -> {
		if (o instanceof Boolean) return (Boolean) o;
		return true;
	};

	public final <R> Mono<R> ifThen(Mono<R> thenMono) {
		return ifThen(thenMono, null);
	}

	public final <R> Mono<R> ifThen(Mono<R> thenMono, @Nullable Mono<R> elseMono) {
		return ifThen(BOOLEAN_OR_THERE_PREDICATE, thenMono, elseMono);
	}

	public final <R> Mono<R> ifThen(Predicate<? super T> ifPredicate,
			Mono<R> thenMono) {
		return ifThen(ifPredicate, thenMono, null);
	}

	public final <R> Mono<R> ifThen(Predicate<? super T> ifPredicate,
			Mono<R> thenMono, @Nullable Mono<R> elseMono) {
		Mono<R> valuePath = flatMap(v -> {
			if (ifPredicate.test(v)) {
				return thenMono;
			}
			else if (elseMono != null) {
				return elseMono;
			}
			else {
				return Mono.empty();
			}
		});

		if (elseMono != null) {
			return valuePath.switchIfEmpty(elseMono);
		}
		return valuePath;
	}

	public final <R> Flux<R> ifThenMany(Publisher<R> thenPublisher) {
		return ifThenMany(thenPublisher, null);
	}

	public final <R> Flux<R> ifThenMany(Publisher<R> thenPublisher, @Nullable Publisher<R> elsePublisher) {
		return ifThenMany(BOOLEAN_OR_THERE_PREDICATE, thenPublisher, elsePublisher);
	}

	public final <R> Flux<R> ifThenMany(Predicate<? super T> ifPredicate,
			Publisher<R> thenPublisher) {
		return ifThenMany(ifPredicate, thenPublisher, null);
	}

	public final <R> Flux<R> ifThenMany(Predicate<? super T> ifPredicate,
			Publisher<R> thenPublisher, @Nullable Publisher<R> elsePublisher) {
		Flux<R> valuePath = flatMapMany(v -> {
			if (ifPredicate.test(v)) {
				return thenPublisher;
			}
			else if (elsePublisher != null) {
				return elsePublisher;
			}
			else {
				return Mono.empty();
			}
		});

		if (elsePublisher != null) {
			return valuePath.switchIfEmpty(elsePublisher);
		}
		return valuePath;
	}
```

The variant without predicate would detect a `Boolean` value and interpret that as the predicate. Otherwise, if it is not a `Boolean` it would just return `true` and only consider an empty source for the `else` path.

For instance:
```
Mono<Void> act = Flux.range(1, 9)
				.all(i -> i < 10)
				.ifThen(Mono.fromRunnable(() -> System.out.println("did all under 10")));

		act.block(); //prints `did all under 10`
```

**Comment by simonbasle** [comment]
After discussing this, for now on it feels like the added value of such an operator is not high enough. Let's not grow the `Flux` API with something that can already be achieved now (with a little boilerplate).

**Comment by vinaysoni** [comment]
I am not understaing why you concluded that it is not high value.
This has the potential to help reactor achieve the same level of composability as apache camel. If there was a "if" condition available, that could allow inspection of data passing through the flow, one could easily create routes, like it is done in apache camel.
Right now there doesn't seem to be anything like that available. This, for people who come from a camel background is a serious limitation.

**Comment by simonbasle** [comment]
@vinaysoni we felt like introducing such an API for something that could be achieved by a simple `flatMap` was not really worth it, see:

```java
Mono<Void> act = Flux.range(1, 9)
				.all(i -> i < 10)
                .flatMap(allTrue -> allTrue 
                    ? Mono.fromRunnable(() -> System.out.println("did all under 10"))
                    : Mono.empty()
                );
```

**Comment by vinaysoni** [comment]
I agree. That is simple enough. Being new, somehow missed it. Thanks.

**Comment by simonbasle** [comment]
no problem, sometime a tiny bit of good old imperative flow control inside a `flatMap` is all you need 😄

**Comment by ydewit** [comment]
@simonbasle It seems to me that there is still a case for a static `if then else` that is not covered by the above example.

Consider the case where one wants to branch on whether a **Mono is empty or not** and return a Flux:
```java
if( isEmpty() ){
  // doA
  return Flux.empty(); // TODO
} else {
  // doB
  return Flux.empty(); // TODO
}
```
Here we have got a mono from a repository that may or may not be empty and we need to generate a Flux for when the Mono is empty and another completely different Flux when the Mono is not empty.

How to achieve that?

A `mono.switchIfEmpty().flatMapMany()` is the equivalent to `if( isEmpty() ) getDefault(); doSomething();`.

The following will not work in all cases. Consider when the `doA` branch happens to return an empty flux: in this case we do not want to execute `doB`:
```java
return find(cr.getKey()).flatMapMany(mr -> {
            // doA
            return Flux.empty();
        }).switchIfEmpty(Flux.defer(() -> {
            // doB
            return Flux.empty();
        }));
```

**Comment by ydewit** [comment]
On Gitter, the work around Leonard Brunnings came up with was to emit a special `empty` value. Considering that I would also have to remove this `empty` value to return a clean Flux to the caller, it is a less than optimal workaround to say the least.

**Comment by simonbasle** [comment]
@ydewit I'm not sure I get why the "empty object" needs to be filtered out?

```java
dbMono
  .returnIfEmpty(EMPTY_SPECIAL_INSTANCE)
  .flatMapMany(v -> {
    if (v == EMPTY_SPECIAL_INSTANCE) return fluxDoA(); //can still be empty
    return fluxDoB(); //can even depend on the value v from the db
  });
```

**Comment by ydewit** [comment]
Ha! I now see it in plain colors!

thanks

**Comment by 391861737** [comment]
@vinaysoni It worked! Nice job.

**Comment by Cryptoevil** [comment]
If you need only `Mono`, try out this:
`Mono.just(0)
                .filter(condition -> condition != 1)
                .switchIfEmpty(Mono.just(1));`

**Comment by botu** [comment]
And if I want to save data but not to return Mono<Void> ? 
Something like this
(clientResponse)->
{
if (clientResponse.statusCode.isError())
{
    return Mono.error(...)
}
return clientReponse;
}

I can't use filter with true/false (at first I try but now I realise that I need the information about the error on next step and filter will never give it to me)

I can't use .map or .flatMap because I am returning different types of data.
Of course, I can skip it to catch and work with the error on the next step, but maybe there is some other solution to return Mono.error or Mono<...> depending of the input data ?

**Comment by simonbasle** [comment]
> I can't use .map or .flatMap because I am returning different types of data.

nothing prevents you from returning the same type of data.

Alternatively you could even technically throw from either `map`, `filter` or even `doOnNext`:

```java
responseFlux.doOnNext(clientResponse -> {
    if (clientResponse.statusCode.isError()) {
        throw new WhateverException();
    }
});
```

**Comment by robotmrv** [comment]
as alternative `reactor.core.publisher.Mono#delayUntil` could be used.   

Regarding throwing error while operating with `ClientResponse`. It could cause memory leaks if response body was not consumed.  
see warning from https://docs.spring.io/spring-framework/docs/5.2.x/spring-framework-reference/web-reactive.html#webflux-client-exchange

**Comment by ataronet** [comment]
Is this function `ifThen` available in Mono? I can't see it in the documentation.

**Comment by sbilello** [comment]
@simonbasle Is there any idea to introduce such `ifThenElse` logic?

**Comment by simonbasle** [comment]
@sbilello please read the whole issue and understand that the idea of adding such an operator has been dropped

**Comment by 62mkv** [comment]
@simonbasle I've read the whole issue but still don't quite follow, why exactly is something like `switchIfEmptyOrElse` is not introduced. 

For example we have a code like this:
```Java
return getPlayerIdAsMono()
     .flatMapMany(player -> getGamesListForPlayerAsFlux(player))
     .switchIfEmpty(getGamesListWithoutPlayerAsFlux())
     .flatMap(Flux::fromIterable);
```

This will return a `Flux<Game>` for player, _if we have a player_ otherwise it will return Flux for some general collection of games. 

But in order to achieve that we have to fall back to this trick with returning a `Flux<List<Game>>` from `getGamesListForPlayerAsFlux` instead of returning just `Flux<Game>`, as there is otherwise no way to distinguish if we have a _blank collection of games per player_ OR _a missing player (information)_

Having something akin to 
```Java
return getPlayerIdAsMono()
     .switchIfEmptyOrElse(
          getGamesWithoutPlayerAsFlux(), 
          player -> Mono.just(player).flatMapMany(player -> getGamesForPlayerAsFlux(player));
```

In other words, Reactor has an idiom of `if source empty then A` but seems to miss `is source empty then A else B`. Falling back to returning "special values" as suggested in the thread above, or a special type, as we do in the existing solution (described hereabove), feels second-class...

**Comment by 62mkv** [comment]
I also do not feel like https://github.com/reactor/reactor-core/pull/1187 is really about the same thing, because it would apply branching logic based on a condition over _emitted value_ which is trivial to achieve with `.flatMap`, whereas what's missing is a capability to apply proper branching (with 2 branches, not just 1) based on a condition over `Mono` completion state (if it had emitted anything or not). 

Maybe something like `flatMapAny` would help, that would present empty Mono as a `null` value to a `Function`, provided as an argument.

**Comment by DidierLoiseau** [comment]
Oh, so this is why my colleague wrote
```java
.map(Optional::of)
.defaultIfEmpty(Optional.empty())
.flatMap(/* do A or B depending on the Optional emptyness */)
```
I also found [a similar solution on SO](https://stackoverflow.com/a/66879026/525036).

In our case, we need to return a `Mono<Void>` as this is for WebFlux, so as a workaround, this is also possible:
```java
.switchIfEmpty(Mono.defer(() -> {
	// do stuff when empty
	return exchange.getResponse()
			.writeWith(…)
			.then(Mono.empty());
}))
.flatMap(value -> {
	// do stuff when non-empty
});
```
but this trick might not be clear to a future developer, and it wouldn’t work if `switchIfEmpty()` needs to return a non-empty `Mono`.

I agree it is not the same issue as the original one but I couldn’t find another issue for it, and since I am new to Reactor & WebFlux, I might just be completely overlooking something.
